File: recipes/claude/28-asynchronous-batch-processing/typescript/src/batch_processor.ts
import Anthropic from '@anthropic-ai/sdk';
import * as dotenv from 'dotenv';

dotenv.config();

const apiKey = process.env.ANTHROPIC_API_KEY;
if (!apiKey) {
  console.error('ERROR: ANTHROPIC_API_KEY environment variable is required.');
  process.exit(1);
}

const client = new Anthropic({ apiKey });

interface BatchPrompt {
  id: string;
  prompt: string;
}

interface ParsedBatchResult {
  customId: string;
  status: 'succeeded' | 'errored' | 'canceled' | 'expired';
  content?: string;
  error?: string;
  inputTokens?: number;
  outputTokens?: number;
}

async function createBatch(prompts: BatchPrompt[]) {
  console.log(`[+] Submitting batch containing ${prompts.length} requests...`);
  
  const requests = prompts.map((item) => ({
    custom_id: item.id,
    params: {
      model: 'claude-3-5-sonnet-20241022',
      max_tokens: 500,
      messages: [{ role: 'user' as const, content: item.prompt }]
    }
  }));

  const batch = await client.messages.batches.create({ requests });
  console.log(`[+] Batch created successfully! ID: ${batch.id}`);
  console.log(`    Processing status: ${batch.processing_status}`);
  return batch;
}

async function pollBatch(batchId: string, intervalSeconds: number = 5, timeoutSeconds: number = 300) {
  const startTime = Date.now();
  console.log(`[+] Polling batch status for ${batchId} every ${intervalSeconds}s...`);

  while (true) {
    const batch = await client.messages.batches.retrieve(batchId);
    const counts = batch.request_counts;
    console.log(
      `    Status: ${batch.processing_status} | ` +
      `Processing: ${counts.processing} | ` +
      `Succeeded: ${counts.succeeded} | ` +
      `Errored: ${counts.errored} | ` +
      `Canceled: ${counts.canceled} | ` +
      `Expired: ${counts.expired}`
    );

    if (batch.processing_status === 'ended') {
      console.log(`[+] Batch ended at ${batch.ended_at}`);
      return batch;
    }

    if ((Date.now() - startTime) / 1000 > timeoutSeconds) {
      console.log(`[!] Polling timed out after ${timeoutSeconds}s.`);
      return batch;
    }

    await new Promise((resolve) => setTimeout(resolve, intervalSeconds * 1000));
  }
}

async function streamResults(batchId: string): Promise<Map<string, ParsedBatchResult>> {
  console.log(`[+] Streaming results for batch ${batchId}...`);
  const results = new Map<string, ParsedBatchResult>();

  const resultsStream = await client.messages.batches.results(batchId);

  for await (const entry of resultsStream) {
    const customId = entry.custom_id;
    const resultType = entry.result.type;

    if (resultType === 'succeeded') {
      const msg = entry.result.message;
      const text = msg.content
        .filter((c) => c.type === 'text')
        .map((c) => (c as { type: 'text'; text: string }).text)
        .join('\n');

      results.set(customId, {
        customId,
        status: 'succeeded',
        content: text,
        inputTokens: msg.usage.input_tokens,
        outputTokens: msg.usage.output_tokens
      });
      console.log(`  [${customId}] SUCCEEDED (${msg.usage.input_tokens} in / ${msg.usage.output_tokens} out tokens)`);
    } else if (resultType === 'errored') {
      const err = entry.result.error;
      results.set(customId, {
        customId,
        status: 'errored',
        error: JSON.stringify(err)
      });
      console.log(`  [${customId}] ERRORED:`, err);
    } else if (resultType === 'canceled') {
      results.set(customId, { customId, status: 'canceled' });
      console.log(`  [${customId}] CANCELED`);
    } else if (resultType === 'expired') {
      results.set(customId, { customId, status: 'expired' });
      console.log(`  [${customId}] EXPIRED`);
    }
  }

  return results;
}

async function cancelBatch(batchId: string) {
  console.log(`[!] Requesting cancellation for batch ${batchId}...`);
  const batch = await client.messages.batches.cancel(batchId);
  console.log(`[!] Batch status: ${batch.processing_status}`);
  return batch;
}

async function main() {
  const samplePrompts: BatchPrompt[] = [
    { id: 'summary-01', prompt: 'Summarize the role of message batches in large data pipelines.' },
    { id: 'summary-02', prompt: 'List 3 architectural patterns for handling 24-hour asynchronous SLAs.' }
  ];

  const batch = await createBatch(samplePrompts);
  const finalBatch = await pollBatch(batch.id, 3, 60);

  if (finalBatch.processing_status === 'ended') {
    const results = await streamResults(batch.id);
    console.log(`\n[+] Finished streaming ${results.size} items.`);
  } else {
    console.log(`\n[*] Batch processing still active or timed out. Batch ID: ${batch.id}`);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

export { createBatch, pollBatch, streamResults, cancelBatch };

File: recipes/claude/28-asynchronous-batch-processing/python/batch_processor.py
"""
Anthropic Message Batches API Client.

Demonstrates submitting asynchronous batches with custom_id mapping,
monitoring batch lifecycle status, handling errors, streaming JSONL results,
and batch cancellation.
"""

import os
import sys
import time
from typing import Dict, Any, List, Optional
from anthropic import Anthropic
from anthropic.types.message_create_params import MessageCreateParamsNonStreaming
from anthropic.types.messages.batch_create_params import Request


def init_client() -> Anthropic:
    """Initialize Anthropic client using environment variable."""
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        print("ERROR: ANTHROPIC_API_KEY environment variable is missing.", file=sys.stderr)
        sys.exit(1)
    return Anthropic(api_key=api_key)


def build_batch_requests(prompts: List[Dict[str, str]], model: str = "claude-3-5-sonnet-20241022") -> List[Request]:
    """Construct individual batch requests tagged with deterministic custom_ids."""
    requests: List[Request] = []
    for item in prompts:
        req = Request(
            custom_id=item["id"],
            params=MessageCreateParamsNonStreaming(
                model=model,
                max_tokens=500,
                messages=[
                    {"role": "user", "content": item["prompt"]}
                ]
            )
        )
        requests.append(req)
    return requests


def submit_batch(client: Anthropic, requests: List[Request]):
    """Submit requests batch to the Anthropic Message Batches API endpoint."""
    print(f"[+] Submitting batch containing {len(requests)} requests...")
    batch = client.messages.batches.create(requests=requests)
    print(f"[+] Batch created successfully! ID: {batch.id}")
    print(f"    Processing status: {batch.processing_status}")
    print(f"    Created at: {batch.created_at}")
    return batch


def poll_batch_status(client: Anthropic, batch_id: str, poll_interval_seconds: int = 5, timeout_seconds: int = 300):
    """
    Poll batch status until it reaches 'ended' status or times out.
    Status transitions: in_progress -> canceling (if canceled) -> ended.
    """
    start_time = time.time()
    print(f"[+] Polling batch status for {batch_id} (interval: {poll_interval_seconds}s)...")
    
    while True:
        batch = client.messages.batches.retrieve(batch_id)
        counts = batch.request_counts
        print(
            f"    Status: {batch.processing_status} | "
            f"Processing: {counts.processing} | "
            f"Succeeded: {counts.succeeded} | "
            f"Errored: {counts.errored} | "
            f"Canceled: {counts.canceled} | "
            f"Expired: {counts.expired}"
        )

        if batch.processing_status == "ended":
            print(f"[+] Batch completed with final status 'ended' at {batch.ended_at}")
            return batch

        if time.time() - start_time > timeout_seconds:
            print(f"[!] Polling timed out after {timeout_seconds}s. Batch is still in progress.")
            return batch

        time.sleep(poll_interval_seconds)


def retrieve_and_parse_results(client: Anthropic, batch_id: str) -> Dict[str, Any]:
    """
    Stream and inspect JSONL results from completed batch.
    Maps custom_id to individual results (succeeded, errored, canceled, expired).
    """
    print(f"[+] Streaming batch results for {batch_id}...")
    results_map: Dict[str, Any] = {}
    
    for result in client.messages.batches.results(batch_id):
        cid = result.custom_id
        res_type = result.result.type

        if res_type == "succeeded":
            message = result.result.message
            text_blocks = [b.text for b in message.content if hasattr(b, "text")]
            content_text = "\n".join(text_blocks)
            usage = message.usage
            results_map[cid] = {
                "status": "succeeded",
                "content": content_text,
                "input_tokens": usage.input_tokens,
                "output_tokens": usage.output_tokens,
            }
            print(f"  [{cid}] SUCCEEDED ({usage.input_tokens} in / {usage.output_tokens} out tokens)")
        elif res_type == "errored":
            err = result.result.error
            results_map[cid] = {
                "status": "errored",
                "error": str(err)
            }
            print(f"  [{cid}] ERRORED: {err}")
        elif res_type == "canceled":
            results_map[cid] = {"status": "canceled"}
            print(f"  [{cid}] CANCELED")
        elif res_type == "expired":
            results_map[cid] = {"status": "expired"}
            print(f"  [{cid}] EXPIRED")

    return results_map


def cancel_batch(client: Anthropic, batch_id: str):
    """Cancel an active in-flight batch."""
    print(f"[!] Requesting cancellation for batch {batch_id}...")
    batch = client.messages.batches.cancel(batch_id)
    print(f"[!] Batch status updated to: {batch.processing_status}")
    return batch


def main():
    """Example workflow execution."""
    client = init_client()

    sample_tasks = [
        {"id": "doc-audit-001", "prompt": "Summarize the key benefits of batch processing APIs in 2 sentences."},
        {"id": "doc-audit-002", "prompt": "Extract 3 technical best practices for background async worker design."},
        {"id": "doc-audit-003", "prompt": "Explain why custom_id mapping is critical for idempotency in batch workloads."}
    ]

    requests = build_batch_requests(sample_tasks)
    batch = submit_batch(client, requests)
    final_batch = poll_batch_status(client, batch.id, poll_interval_seconds=3, timeout_seconds=60)

    if final_batch.processing_status == "ended":
        results = retrieve_and_parse_results(client, batch.id)
        print(f"\n[+] Total parsed items: {len(results)}")
    else:
        print(f"\n[*] Batch still processing or timed out. Check batch ID later: {batch.id}")


if __name__ == "__main__":
    main()
